### PR TITLE
Support new IE API

### PR DESCRIPTION
**Pull request by l-bat**
```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2019r2.0:16.04
build_image:Custom Win=openvino-2019r2.0
build_image:Custom Mac=openvino-2019r2.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
test_opencl:Custom=ON
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/src/layers/blank_layer.cpp [hunk 1]
```diff
@@ -111,8 +111,12 @@ class BlankLayerImpl CV_FINAL : public BlankLayer
...
     {
         InferenceEngine::DataPtr input = infEngineDataNode(inputs[0]);
-        CV_Assert(!input->dims.empty());
-
+#if INF_ENGINE_VER_MAJOR_LE(2019010000)
+    CV_Assert(!input->dims.empty());
+#else
+    std::vector<size_t> dims = input->getTensorDesc().getDims();
+    CV_Assert(!dims.empty());
+#endif
```
Comment (dkurt) [inline modules/dnn/src/layers/blank_layer.cpp]: Can we use `getTensorDesc` with older versions? Looks like it was available in 2018R5 and 2019R1 as well.

File: modules/dnn/test/test_ie_models.cpp [hunk 2]
```diff
@@ -114,13 +114,21 @@ static const std::vector<std::string> getOpenVINOTestModelsList()
...
     randu(m, -1, 1);
 
     dataPtr = make_shared_blob<float>(Precision::FP32, dims, (float*)m.data);
+#else
+    m.create(std::vector<int>(dims.begin(), dims.end()), CV_32F);
+    randu(m, -1, 1);
+
+    InferenceEngine::TensorDesc td(Precision::FP32, dims, Layout::ANY);
+    dataPtr = make_shared_blob<float>(td, (float*)m.data);
+#endif
```
Comment (dkurt) [inline modules/dnn/test/test_ie_models.cpp]: `TensorDesc::getDims ` returns reversed dims?

File: modules/dnn/src/op_inf_engine.cpp [hunk 3]
```diff
@@ -233,34 +281,59 @@ static InferenceEngine::DataPtr wrapToInfEngineDataNode(const Mat& m, const std:
...
+#else
+    if (m.type() == CV_32F) {
+        InferenceEngine::TensorDesc td(InferenceEngine::Precision::FP32, shape, layout);
+        return InferenceEngine::make_shared_blob<float>(td, (float*)m.data);
+    }
+    else if (m.type() == CV_8U) {
+        InferenceEngine::TensorDesc td(InferenceEngine::Precision::U8, shape, layout);
+        return InferenceEngine::make_shared_blob<uint8_t>(td, (uint8_t*)m.data);
+    }
+#endif
```
Comment (dkurt) [inline modules/dnn/src/op_inf_engine.cpp]: Can we use `make_shared_blob` with `TensorDesc` for all the supported versions?

File: modules/dnn/test/test_ie_models.cpp (hunk 2 above)
Comment (l-bat) [reply modules/dnn/test/test_ie_models.cpp]: No

Comment (dkurt) [reply modules/dnn/test/test_ie_models.cpp]: I mean there were
```cpp
std::vector<int> reversedDims(dims.begin(), dims.end());
    std::reverse(reversedDims.begin(), reversedDims.end());

    m.create(reversedDims, CV_32F);
    randu(m, -1, 1);
```
Shall we use std::reverse here or not? BTW, if 2018R5 and 2019R1 also support `make_shared_blob` from `TensorDesc` we can just rewrite it.

File: modules/dnn/test/test_ie_models.cpp [hunk 4]
```diff
@@ -114,13 +114,11 @@ static const std::vector<std::string> getOpenVINOTestModelsList()
...
-    std::vector<int> reversedDims(dims.begin(), dims.end());
-    std::reverse(reversedDims.begin(), reversedDims.end());
-
-    m.create(reversedDims, CV_32F);
+    m.create(std::vector<int>(dims.begin(), dims.end()), CV_32F);
     randu(m, -1, 1);
 
-    dataPtr = make_shared_blob<float>(Precision::FP32, dims, (float*)m.data);
+    InferenceEngine::TensorDesc td(Precision::FP32, dims, Layout::ANY);
+    dataPtr = make_shared_blob<float>(td, (float*)m.data);
```
Comment (dkurt) [inline modules/dnn/test/test_ie_models.cpp]: Can we write something like
```cpp
dataPtr = make_shared_blob<float>({Precision::FP32, dims, Layout::ANY}, (float*)m.data);
```
?

File: modules/dnn/src/dnn.cpp [hunk 5]
```diff
@@ -713,21 +713,23 @@ struct DataLayer : public Layer
...
         weights->allocate();
-        weights->set(std::vector<float>(numChannels, scaleFactors[0]));
+
+        float* weight_buf = weights->buffer().as<float*>();
+        std::fill(weight_buf, weight_buf + numChannels, scaleFactors[0]);
 
         // Mean subtraction
-        auto biases = InferenceEngine::make_shared_blob<float>(InferenceEngine::Precision::FP32,
-                                                               {numChannels});
+        auto biases = InferenceEngine::make_shared_blob<float>(td);
```
Comment (dkurt) [inline modules/dnn/src/dnn.cpp]: Biases and weights share the same `TensorDesc`?

File: modules/dnn/src/layers/convolution_layer.cpp [hunk 6]
```diff
@@ -485,9 +484,10 @@ class ConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl
             }
             else
             {
-                ieWeights = InferenceEngine::make_shared_blob<float>(
-                                    InferenceEngine::Precision::FP32, layout,
-                                    ieWeights->dims());
+                ieWeights = InferenceEngine::make_shared_blob<float>({
+                            InferenceEngine::Precision::FP32,
+                            ieWeights->getTensorDesc().getDims(), layout
+                            });
```
Comment (dkurt) [inline modules/dnn/src/layers/convolution_layer.cpp]: Please fix indentation

File: modules/dnn/src/layers/convolution_layer.cpp [hunk 7]
```diff
@@ -1877,9 +1877,10 @@ class DeConvolutionLayerImpl CV_FINAL : public BaseConvolutionLayerImpl
         auto ieWeights = wrapToInfEngineBlob(blobs[0], layout);
         if (fusedWeights)
         {
-            ieWeights = InferenceEngine::make_shared_blob<float>(
-                                InferenceEngine::Precision::FP32, layout,
-                                ieWeights->dims());
+            ieWeights = InferenceEngine::make_shared_blob<float>({
+                        InferenceEngine::Precision::FP32,
+                        ieWeights->getTensorDesc().getDims(), layout
+                        });
```
Comment (dkurt) [inline modules/dnn/src/layers/convolution_layer.cpp]: Please fix indentation here and in other places.

File: modules/dnn/src/layers/concat_layer.cpp [hunk 8]
```diff
@@ -304,7 +304,7 @@ class ConcatLayerImpl CV_FINAL : public ConcatLayer
         InferenceEngine::DataPtr input = infEngineDataNode(inputs[0]);
 
         InferenceEngine::Builder::ConcatLayer ieLayer(name);
-        ieLayer.setAxis(clamp(axis, input->dims.size()));
+        ieLayer.setAxis(clamp(axis, input->getTensorDesc().getDims().size()));
```
Comment (dkurt) [inline modules/dnn/src/layers/concat_layer.cpp]: Can we use just `input->getDims().size()`?

File: modules/dnn/src/op_inf_engine.cpp [hunk 9]
```diff
@@ -45,13 +45,21 @@ infEngineWrappers(const std::vector<Ptr<BackendWrapper> >& ptrs)
...
 }
 
 InfEngineBackendNet::InfEngineBackendNet(InferenceEngine::CNNNetwork& net) : netBuilder(""), cnn(net)
 {
     hasNetOwner = true;
+#if INF_ENGINE_VER_MAJOR_LE(INF_ENGINE_RELEASE_2019R1)
     targetDevice = InferenceEngine::TargetDevice::eCPU;
+#else
+    device_name = "CPU";
+#endif
```
Comment (dkurt) [inline modules/dnn/src/op_inf_engine.cpp]: Can we use just text `device_name` and load plugins for 2018R5 and 2019R1 by name? Using `getPluginByName`.

**Comment by l-bat** [comment]
> @l-bat, there are still number of `getTensorDesc()` which, probably, could be replaced:
> 
> `blob->getTensorDesc().getDims();` -> `blob->getDims();`
> `blob->getTensorDesc().getPrecision();` -> `blob->getPrecision();`
> and so on. Can you please check all of them? (Ctrl+F for "getTensorDesc().")

`error: ‘class InferenceEngine::Blob’ has no member named ‘getDims’`

I have already replaced it wherever possible

File: modules/dnn/src/op_inf_engine.cpp [hunk 10]
```diff
@@ -462,14 +474,25 @@ void InfEngineBackendNet::initPlugin(InferenceEngine::ICNNNetwork& net)
...
 #endif
         }
+#if INF_ENGINE_VER_MAJOR_LE(INF_ENGINE_RELEASE_2019R1)
         plugin = InferenceEngine::InferencePlugin(enginePtr);
-
         netExec = plugin.LoadNetwork(net, {});
+#else
+        std::shared_ptr<InferenceEngine::ICNNNetwork> cnn_net(&net, [] (InferenceEngine::ICNNNetwork*) {});
+        netExec = core.LoadNetwork(InferenceEngine::CNNNetwork(cnn_net), device_name);
```
Comment (dkurt) [inline modules/dnn/src/op_inf_engine.cpp]: Please replace `initPlugin`'s input from ICNNNetwork to CNNNetwork as in here: https://github.com/opencv/opencv/pull/15063/files#diff-da37b66e77aa9907ebc193c732dd00d1R95

It'll help to use just `netExec = core.LoadNetwork(net)`;

File: modules/dnn/src/op_inf_engine.hpp [hunk 11]
```diff
@@ -98,11 +99,16 @@ class InfEngineBackendNet
...
-    InferenceEngine::InferencePlugin plugin;
     InferenceEngine::ExecutableNetwork netExec;
     InferenceEngine::BlobMap allBlobs;
-    InferenceEngine::TargetDevice targetDevice;
+    std::string device_name;
+#if INF_ENGINE_VER_MAJOR_LE(2019010000)
+    InferenceEngine::InferenceEnginePluginPtr enginePtr;
+    InferenceEngine::InferencePlugin plugin;
+#else
+    InferenceEngine::Core core;
```
Comment (dkurt) [inline modules/dnn/src/op_inf_engine.hpp]: If there are two networks withe their own `InferenceEngine::Core` will they share the plugins? I mean can we create two networks in this way on Myriad, in example?

File: modules/dnn/src/op_inf_engine.cpp [hunk 12]
```diff
@@ -361,24 +365,30 @@ static bool detectMyriadX_()
...
+            enginePtr = dispatcher.getPluginByDevice("MYRIAD");
+            sharedPlugins["MYRIAD"] = enginePtr;
         }
     }
     auto plugin = InferenceEngine::InferencePlugin(enginePtr);
     try
     {
         auto netExec = plugin.LoadNetwork(cnn, {{"VPU_PLATFORM", "VPU_2480"}});
+#else
+    InferenceEngine::Core ie;
```
Comment (dkurt) [inline modules/dnn/src/op_inf_engine.cpp]: `getCore`?

File: modules/dnn/src/op_inf_engine.cpp [hunk 13]
```diff
@@ -448,7 +460,12 @@ void InfEngineBackendNet::initPlugin(InferenceEngine::ICNNNetwork& net)
                 {
                     InferenceEngine::IExtensionPtr extension =
                         InferenceEngine::make_so_pointer<InferenceEngine::IExtension>(libName);
+
+#if INF_ENGINE_VER_MAJOR_LE(INF_ENGINE_RELEASE_2019R1)
                     enginePtr->AddExtension(extension, 0);
+#else
+                    getCore().AddExtension(extension, device_name);
+#endif
```
Comment (alalek) [inline modules/dnn/src/op_inf_engine.cpp]: There is extra message on [CI](http://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/2475/steps/test_dnn-opencl/logs/stdio):
> [ WARN:0] DNN-IE: Can't load extension plugin

Perhaps `catch(...) {}` below should be rewritten to capture more information.

File: modules/dnn/test/test_misc.cpp [hunk 14]
```diff
@@ -463,6 +463,39 @@ TEST_P(Async, set_and_forward_all)
     }
 }
 
+typedef testing::TestWithParam<tuple<int, Target> > Async;
+TEST_P(Async, forward_two_nets)
```
Comment (dkurt) [inline modules/dnn/test/test_misc.cpp]: Perhaps, this test should have separate scope due there is no Asynchronous mode is tested.

File: modules/dnn/src/layers/slice_layer.cpp [hunk 15]
```diff
@@ -301,14 +301,14 @@ class SliceLayerImpl : public SliceLayer
         {
             std::vector<size_t> outShape(numDims);
             for (int i = 0; i < numDims; ++i)
-                outShape[numDims - 1 - i] = sliceRanges[0][i].size();
+                outShape[i] = sliceRanges[0][i].size();
 
             ieLayer.getInputPorts()[1].setParameter("type", "weights");
 
-            // Fake blob which will be moved to inputs (as weights).
```
Comment (dkurt) [inline modules/dnn/src/layers/slice_layer.cpp]: Please keep this comment

File: modules/dnn/src/op_inf_engine.cpp (hunk 13 above)
Comment (dkurt) [reply modules/dnn/src/op_inf_engine.cpp]: Fixed by replacing `device_name` which can be both `CPU` and `FPGA` to just `"CPU"`.

**Pull request merged**